### PR TITLE
RT3720 Increment session miss counter properly

### PR DESCRIPTION
**Pull request by kaduk**
Updated version of the patch linked from RT 3720, after the state machine rewrite.

### CODE REVIEW

File: ssl/statem/statem_clnt.c [hunk 1]
```diff
@@ -1122,6 +1122,9 @@ MSG_PROCESS_RETURN tls_process_server_hello(SSL *s, PACKET *pkt)
          * overwritten if the server refuses resumption.
          */
         if (s->session->session_id_length > 0) {
+#ifndef OPENSSL_NO_AKAMAI
```
Comment (ghedo) [inline ssl/statem/statem_clnt.c]: Don't think this is required, no? A lot of these weird defines aren't even documented so we might as well drop them.

Comment (kaduk) [inline ssl/statem/statem_clnt.c]: Indeed, that's a bad cherry-pick, let me fix.

**Comment by kroeckx** [comment]
+1

**Comment by t-j-h** [comment]
I don't think any version of OpenSSL has ever tracked the hit/miss counter on the client side - at least a quick check didn't show that being done. Is there a version of OpenSSL that you know did this?

**Comment by kroeckx** [comment]
I had the same question, but I don't see problem adding it.  I guess they create enough client connections to care.  The server counter seems to be updated correctly.

File: ssl/statem/statem_clnt.c [hunk 2]
```diff
@@ -1122,6 +1122,7 @@ MSG_PROCESS_RETURN tls_process_server_hello(SSL *s, PACKET *pkt)
          * overwritten if the server refuses resumption.
          */
         if (s->session->session_id_length > 0) {
+            s->ctx->stats.sess_miss++;
```
Comment (mattcaswell) [inline ssl/statem/statem_clnt.c]: Is there a race condition problem with this?

**Comment by mattcaswell** [comment]
Ping @kaduk - there is an outstanding question on this.

File: ssl/statem/statem_clnt.c (hunk 2 above)
Comment (kaduk) [inline ssl/statem/statem_clnt.c]: While I'm not super-familiar with this code, ssl_sess.c is already accessing session_id_length and sess_miss with no immediately visible locking.  (Presumably the larger-scope locks like SSL_CTX are not going to be in use here...)

**Comment by mattcaswell** [comment]
Ok, I looked at this. It does look like this could be an issue, but you're right there is no visible locking on the other stats. I guess it's not a huge problem, it just means that the stats may not be 100% reliable.

**Comment by mattcaswell** [comment]
+1

**Comment by mattcaswell** [comment]
Merged.

**Pull request closed without merging**